File: app/src/main/java/com/example/all_sensors/presentation/HrListener.java
package com.example.all_sensors.presentation;

import android.util.Log;

import com.example.all_sensors.presentation.data.DataBuffer;
import com.samsung.android.service.health.tracking.HealthTracker;
import com.samsung.android.service.health.tracking.data.DataPoint;
import com.samsung.android.service.health.tracking.data.ValueKey;

import java.util.List;
import java.util.Locale;

/**
 * Listener para frecuencia cardiaca continua.
 */
public class HrListener implements HealthTracker.TrackerEventListener {
    private final DataBuffer buffer;

    public HrListener(DataBuffer buffer) {
        this.buffer = buffer;
    }

    @Override
    public void onDataReceived(List<DataPoint> dataPoints) {
        long now = System.currentTimeMillis();
        // DataPoint tendrá ValueKey.HeartRateSet.HEART_RATE a 1Hz
        for (com.samsung.android.service.health.tracking.data.DataPoint dp : dataPoints) {
            int hr = dp.getValue(ValueKey.HeartRateSet.HEART_RATE);

            Log.d("ACC_DATA",
                    String.format(Locale.US, "%d", hr));

            buffer.add(new com.example.all_sensors.presentation.data.DataPoint(now,
                    "HR",String.format(Locale.US, "%d", hr)));

            ObserverUpdater.getObserverUpdater().notifyHrObservers(hr);
        }
    }

    @Override
    public void onFlushCompleted() {
        // No suele usarse
    }

    @Override
    public void onError(com.samsung.android.service.health.tracking.HealthTracker.TrackerError error) {
        // TODO: manejo de error
        Log.e("HR_DATA", "HR error");
    }
}

File: app/src/main/java/com/signals/smartwatch/sensors/samsung/HrListener.kt
package com.signals.smartwatch.sensors.samsung

import android.util.Log
import com.signals.smartwatch.SensorEvents
import com.signals.smartwatch.data.SensorDao
import com.signals.smartwatch.data.SensorEntity
import com.samsung.android.service.health.tracking.HealthTracker
import com.samsung.android.service.health.tracking.data.DataPoint
import com.samsung.android.service.health.tracking.data.ValueKey

class HrListener(
    private val dao: SensorDao,
    private val events: SensorEvents
) : HealthTracker.TrackerEventListener {

    override fun onDataReceived(dataPoints: List<DataPoint>) {
        val now = System.currentTimeMillis()
        dataPoints.forEach { dp ->
            val status = dp.getValue(ValueKey.HeartRateSet.HEART_RATE_STATUS)

            // Status 1 = medición válida según documentación Samsung Health SDK
            if (status != STATUS_MEASUREMENT_VALID) {
                Log.w(TAG, "HR descartado — status=$status")
                return@forEach
            }

            val hr = dp.getValue(ValueKey.HeartRateSet.HEART_RATE)
            Log.d(TAG, hr.toString())
            dao.insertBlocking(SensorEntity(timestamp = now, sensor = "HR", values = hr.toString()))
            events.onHr(hr)
        }
    }

    override fun onFlushCompleted() = Unit

    override fun onError(error: HealthTracker.TrackerError) {
        Log.e(TAG, "Error: $error")
    }

    companion object {
        private const val TAG = "HR_DATA"
        private const val STATUS_MEASUREMENT_VALID = 1
    }
}
